File: ImperatorToCK3/Source/Mappers/ReligionMapper/ReligionMapper.cpp
#include "ReligionMapper.h"
#include "Log.h"
#include "ParserHelpers.h"

mappers::ReligionMapper::ReligionMapper()
{
	LOG(LogLevel::Info) << "-> Parsing religion mappings.";
	registerKeys();
	parseFile("configurables/religion_map.txt");
	clearRegisteredKeywords();
	LOG(LogLevel::Info) << "<> Loaded " << religionMappings.size() << " religious links.";
}

mappers::ReligionMapper::ReligionMapper(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void mappers::ReligionMapper::registerKeys()
{
	registerKeyword("link", [this](const std::string& unused, std::istream& theStream) {
		religionMappings.emplace_back(theStream);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

void mappers::ReligionMapper::loadRegionMappers(std::shared_ptr<mappers::ImperatorRegionMapper> impRegionMapper, std::shared_ptr<mappers::CK3RegionMapper> _ck3RegionMapper)
{
	const auto imperatorRegionMapper = std::move(impRegionMapper);
	const auto ck3RegionMapper = std::move(_ck3RegionMapper);
	if (!imperatorRegionMapper)
		throw std::runtime_error("Religion Mapper: Imperator Region Mapper is unloaded!");
	if (!ck3RegionMapper)
		throw std::runtime_error("Religion Mapper: CK3 Region Mapper is unloaded!");
	for (auto& mapping : religionMappings)
	{
		mapping.insertImperatorRegionMapper(imperatorRegionMapper);
		mapping.insertCK3RegionMapper(ck3RegionMapper);
	}
}

std::optional<std::string> mappers::ReligionMapper::match(const std::string& impReligion, const unsigned long long ck3ProvinceID, const unsigned long long impProvinceID) const
{
	for (const auto& religionMapping : religionMappings)
	{
		const auto& possibleMatch = religionMapping.religionMatch(impReligion, ck3ProvinceID, impProvinceID);
		if (possibleMatch)
			return *possibleMatch;
	}
	return std::nullopt;
}

File: ImperatorToCK3/Mappers/Religion/ReligionMapper.cs
using commonItems;
using commonItems.Collections;
using ImperatorToCK3.CK3.Religions;
using ImperatorToCK3.Mappers.Region;
using System;
using System.Collections.Generic;
using System.IO;

namespace ImperatorToCK3.Mappers.Religion;

internal sealed class ReligionMapper {
	public ReligionMapper(ReligionCollection ck3Religions, ImperatorRegionMapper imperatorRegionMapper, CK3RegionMapper ck3RegionMapper) {
		this.imperatorRegionMapper = imperatorRegionMapper;
		this.ck3RegionMapper = ck3RegionMapper;

		Logger.Info("Parsing religion mappings...");

		var parser = new Parser();
		RegisterKeys(parser);
		var mappingsPath = Path.Combine("configurables", "religion_map.txt");
		parser.ParseFile(mappingsPath);

		Logger.Info($"Loaded {religionMappings.Count} religious links.");

		RemoveMappingsWithNonexistentCK3Faiths(ck3Religions);
		BuildLookup();

		Logger.IncrementProgress();
	}
	public ReligionMapper(BufferedReader reader, ReligionCollection ck3Religions, ImperatorRegionMapper imperatorRegionMapper, CK3RegionMapper ck3RegionMapper) {
		this.imperatorRegionMapper = imperatorRegionMapper;
		this.ck3RegionMapper = ck3RegionMapper;

		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);

		RemoveMappingsWithNonexistentCK3Faiths(ck3Religions);
		BuildLookup();
	}

	private void RemoveMappingsWithNonexistentCK3Faiths(ReligionCollection ck3Religions) {
		religionMappings.RemoveWhere(m=>m.CK3FaithId is not null && ck3Religions.GetFaith(m.CK3FaithId) is null);
	}

	private void BuildLookup() {
		religionLookup.Clear();
		foreach (var mapping in religionMappings) {
			foreach (var irReligionId in mapping.IrReligionIds) {
				if (!religionLookup.TryGetValue(irReligionId, out var list)) {
					list = new List<ReligionMapping>();
					religionLookup[irReligionId] = list;
				}
				list.Add(mapping);
			}
		}
	}

	public string? Match(string irReligionId, string? ck3CultureId, ulong? ck3ProvinceId, ulong? irProvinceId, string? irHistoricalTag, Configuration config) {
		if (string.IsNullOrEmpty(irReligionId)) {
			return null;
		}

		if (!religionLookup.TryGetValue(irReligionId, out var candidates)) {
			return null;
		}

		foreach (var religionMapping in candidates) {
			var possibleMatch = religionMapping.Match(irReligionId, ck3CultureId, ck3ProvinceId, irProvinceId, irHistoricalTag, config, imperatorRegionMapper, ck3RegionMapper);
			if (possibleMatch is not null) {
				return possibleMatch;
			}
		}
		return null;
	}

	private void RegisterKeys(Parser parser) {
		parser.RegisterKeyword("link", reader => {
			religionMappings.Add(ReligionMapping.Parse(reader));
		});
		parser.IgnoreAndLogUnregisteredItems();
	}
	private readonly List<ReligionMapping> religionMappings = new();
	private readonly Dictionary<string, List<ReligionMapping>> religionLookup = new(StringComparer.Ordinal);
	private readonly ImperatorRegionMapper imperatorRegionMapper;
	private readonly CK3RegionMapper ck3RegionMapper;
}